File: app/adapters/page.rb
class Page
  def self.from(object, options = {})
    "Page::#{object.class.name}".constantize.new(object, options)
  end

  def self.home
    Page::Home.new(nil)
  end

  class Base
    attr_reader :object, :options

    def initialize(object, options = {})
      @object = object
      @options = options
    end

    def serialize
      @serialize ||= PageSerializer.serialize(self)
    end

    def key
      @key ||= params.values.flatten.join('/')
    end

    def params
      raise NotImplementedError
    end

    def props
      raise NotImplementedError
    end

    def schema_data
      return @schema_data if defined?(@schema_data)

      @schema_data = {
        '@context': 'https://schema.org',
        '@type': object.is_snippet? ? 'TechArticle' : 'ItemList',
        url: object.full_url,
        main_entity_of_page: {
          '@type': 'WebPage',
          '@id': object.full_url
        },
      }

      if defined?(additional_schema_data)
        @schema_data.merge!(additional_schema_data)
      end

      @schema_data.deep_transform_keys! do |key|
        key.to_s.camelize(:lower)
      end
    end

    protected

    def slug_segments
      @slug_segments ||= object.slug.slice(1..-1).split('/')
    end
  end
end

File: src/adapters/page.js
import PageSerializer from '#src/serializers/pageSerializer.js';

export default class Page {
  static pageTypes = {};

  // Registers different page types so they can be dynamically created later
  static register(pageType) {
    this.pageTypes[pageType.name] = pageType;
  }

  // Creates a page instance dynamically based on the object type
  static from(object, options = {}) {
    const pageName = `${object.constructor.name}Page`;
    return new this.pageTypes[pageName](object, options);
  }

  static get home() {
    return new this.pageTypes.HomePage();
  }

  constructor(object, options = {}) {
    this.object = object;
    this.options = options;
  }

  get serialize() {
    return PageSerializer.serialize(this);
  }

  get key() {
    return Object.values(this.params).flat().join('/');
  }

  get params() {
    throw new Error('Not implemented');
  }

  get props() {
    throw new Error('Not implemented');
  }

  get schemaData() {
    const schemaData = {
      '@context': 'https://schema.org',
      '@type': this.object.isSnippet ? 'TechArticle' : 'ItemList',
      url: this.object.fullUrl,
      mainEntityOfPage: { '@type': 'WebPage', '@id': this.object.fullUrl },
    };

    if (this.additionalSchemaData) {
      Object.assign(schemaData, this.additionalSchemaData);
    }

    return schemaData;
  }

  get slugSegments() {
    return this.object.slug.slice(1).split('/');
  }
}
